### ISSUE TITLE
Issues opening large files

### ISSUE THREAD

**Issue by Rikk** [notepad-plus-plus/notepad-plus-plus#681]
There seems to be an inconstant behavior regarding opening large files.

Refs:
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/668#issuecomment-128669574
https://notepad-plus-plus.org/community/topic/237/editing-600-mega-xml-file

This is our pet error message for this issue:
![too-big](https://cloud.githubusercontent.com/assets/2916485/9144947/0f44481e-3d27-11e5-81b9-b32b1f52b26c.png)

Some new investigation I made with a clean 6.8.1 (no plugins):
(results today are better than previously posted)
- I can open a single plain-text file with size 575MB, but not one with 600MB. So, today this is my limit.
- Opening the 575MB file, peak working set of RAM is 1.800.000KB (~1.7GB), but after the file is open, working set establishes at 1.210.000KB.
- Trying to open a second file >500MB, is a no go. Npp doesn't even try to open it, not a single kb of memory increases, and our pet error is directly displayed.
- Sometimes, it doesn't even try to open a file (shows error), when I close first file, then try to open it again. This is sporadic.

Using my working copy, with plugins, I can open one of 150MB, but not two -- Npp doesn't even try to; yet it uses only 320MB memory.

**Comment by ariccio** [comment]
Xref: #317

**Comment by milipili** [comment]
http://docs.notepad-plus-plus.org/index.php/Large_Text_Files.

**Comment by informatorius** [comment]
The Large_Text_Files document says it is a memory limit because of using Scintilla takes much ram.
1) For a 64bit Windows this value should be doubled to 4GB and Notepad++ compiled with LargeAdressAware flag.
2) Can Notepad++ or Scintilla code use some RAM compression for the text? 
Todays PC have more than one CPU core and can use that cpu power to save ram using compression.

**Comment by informatorius** [comment]
For 1) I have a easy workaround: Use a program like LargeAdressAware and call it on notepad++.exe then notepad++ can open larger documents.
=> This should be the default, it is only a linker flag

**Comment by chcg** [comment]
See https://stackoverflow.com/questions/2288728/drawbacks-of-using-largeaddressaware-for-32-bit-windows-executables
, about the issue that need to be checked

**Comment by chcg** [comment]
See http://www.scintilla.org/ScintillaHistory.html 3.6.0 feature:
- Report error when attempt to resize buffer to more than 2GB with SC_STATUS_FAILURE.

**Comment by informatorius** [comment]
I will test next notepad++ version with LargeAdressAware when scintilla 3.6.0 is used.

**Comment by informatorius** [comment]
@chcg: I tried new scintilla 3.6.1 and there is no difference.

The LargeAdressAware flag enables Notepad++ also to open more big files in parallel,
because there memory usage adds up until max RAM usage is reached.

So if already a big file is open you may get the "file size limit" dialog if open another big file.
If you close the first big file first then the second big file can be opened successfully.
With LargeAdressAware you can open both big files in parallel.

**Comment by lucasunruh** [comment]
Hi Developers,
I'm working on a school project in a data structures class and my group is looking for a development issue in Notepad++ that would be in some way applicable to the subject of data structures. This issue caught our eye because it was one of the few issues that deals with large memory structures that have the potential for optimization.  At this point we're primarily just exploring, any advice would be greatly appreciated.

**Comment by informatorius** [comment]
@lucasunruh: Notepad++ uses library scintilla. You can try to optimize large memory data structures in scintilla.

**Comment by arhangelsoft** [comment]
Hello developers,

why you don't use File mapping methods from windows API for reading a huge files?

**Comment by MarkusAO** [comment]
For those of you who can open large files, but find that browsing / editing them is painfully difficult and often freezes NP++ up... After the file opens up, switch off syntax highlighting. E.g. select a light-weight syntax from the Language menu (.ini?), or append .txt to your file for the duration.

**Comment by arhangelsoft** [comment]
@MarkusAO I think it's an idea for pull request. Check that the opening file it's a large (option to define "large" file size make editable trogth options) and force scintilla style to txt file.

**Comment by GokouZWAR** [comment]
Just tried to open a 1.5GB text file in notepad++ and it didn't open it. I have 12gb of memory on this system with 8gbs free so can we please cut the crap and answer the question: "How big of a file can I open?"

What part of the code says "This file is too big because its X size, throw error <you cannot open this>"? I just want to know what X is.

This will help me know how big I can have the file and how many times I'll need to split it to be ABLE to open it with notepad++. Thanks.

**Comment by garydgregory** [comment]
It looks like Scintilla 4.0.5 has experimental support for file larger than 2 GB.

**Comment by LtqxWYEG** [comment]
This is bad: I wanted to open a 953 MB (1.000.000.002 Bytes) big file and Notepad++ uses 7.84 GB RAM, very slowly dropping. I have 16 GB of RAM, so it isn't an issue; but what if I wanted to open a file that is 2 GiB big?
(I am opening files with digits of PI)

Why does it have to multiply the data 8 times? Is there a program that handles big text files flawlessly? (Meaning, only 1 GiB RAM when file is 1 GiB big)

Yes, I used the GB and GiB correctly :P

**Comment by MetaChuh** [comment]
the memory consumption and fluctuation is caused by the scintilla component as well as some plugins.
please remove plugins like DSpellCheck, or other active text/lexer plugins if you are using large files, especially files with very long lines without line breaks.

regarding scintilla: unfortunately scintilla is an external project, so you will have to file your issue at the [<kbd>scintilla bug tracker</kbd>](https://sourceforge.net/p/scintilla/bugs/).
there is currently nothing we can currently do about this experience, at the notepad++ part.
we apologise for that.

-----
```
general notes:

when creating a new issue, please provide all information as shown at the issue template.
your notepad++ debug info is part of the minimum requirements.
(? > Debug Info... > Copy debug info into clipboard).

please also conduct a search at the issue tracker beforehand, to check if a similar issue
already exists.

for general questions, or if it is not sure, whether your issue is directly related to
the notepad++ source code, please visit us at the notepad++ community forum, and
search if related topics exists. you are welcome to post either at similar
topics, or to create a new topic at Help Wanted or General Discussion.
(no extra account is needed, just use your github account to sign in)
```
[<p align="center"><kbd> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; Click here to visit the Notepad++ Community Forum &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;&emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; </kbd></p>](https://notepad-plus-plus.org/community/)
